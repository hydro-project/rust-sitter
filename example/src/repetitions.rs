#[rust_sitter::grammar("repetitions")]
pub mod grammar {
    use rust_sitter::Spanned;

    #[rust_sitter::language]
    #[derive(Debug)]
    #[allow(dead_code)]
    pub struct NumberList {
        #[rust_sitter::repeat(non_empty = true)]
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
        numbers: Spanned<Vec<Spanned<i32>>>,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }
}

#[rust_sitter::grammar("repetitions_without_delim")]
pub mod grammar2 {
    use rust_sitter::Spanned;

    #[rust_sitter::language]
    #[derive(Debug)]
    #[allow(dead_code)]
    pub struct NumberList {
        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
        numbers: Spanned<Vec<Spanned<i32>>>,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }
}

#[rust_sitter::grammar("repetitions_optional_elem")]
pub mod grammar3 {
    use rust_sitter::Spanned;

    #[rust_sitter::language]
    #[derive(Debug)]
    #[allow(dead_code)]
    pub struct NumberList {
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
        numbers: Spanned<Vec<Spanned<Option<i32>>>>,
        #[rust_sitter::skip(123)]
        metadata: u32,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn repetitions_grammar() {
        insta::assert_debug_snapshot!(grammar::parse(""));
        insta::assert_debug_snapshot!(grammar::parse("1"));
        insta::assert_debug_snapshot!(grammar::parse("1, 2"));
    }

    #[test]
    fn repetitions_grammar2() {
        insta::assert_debug_snapshot!(grammar2::parse(""));
        insta::assert_debug_snapshot!(grammar2::parse("1"));
        insta::assert_debug_snapshot!(grammar2::parse("1 2"));
    }

    #[test]
    fn repetitions_grammar3() {
        insta::assert_debug_snapshot!(grammar3::parse(""));
        insta::assert_debug_snapshot!(grammar3::parse("1,"));
        insta::assert_debug_snapshot!(grammar3::parse("1, 2"));
        insta::assert_debug_snapshot!(grammar3::parse("1,, 2"));
        insta::assert_debug_snapshot!(grammar3::parse("1,, 2,"));
    }
}
