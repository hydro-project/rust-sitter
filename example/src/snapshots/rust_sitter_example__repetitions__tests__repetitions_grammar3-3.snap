---
source: example/src/repetitions.rs
expression: "grammar3::parse(\"1, 2\")"
---
Ok(
    NumberList {
        numbers: Spanned {
            value: [
                Spanned {
                    value: Some(
                        1,
                    ),
                    span: (
                        0,
                        1,
                    ),
                },
                Spanned {
                    value: Some(
                        2,
                    ),
                    span: (
                        3,
                        4,
                    ),
                },
            ],
            span: (
                0,
                4,
            ),
        },
        metadata: 123,
    },
)
